### ISSUE TITLE
New dependency (hypothesis) creates a long new build dependency chain

### ISSUE THREAD

**Issue by npmccallum** [pyca/cryptography#2666]
Even if you exclude all the optional (extra) dependencies, distributions which strive to run tests on all packages which provide them will be forced to package a large number of packages under the flake8 test dependency of python-hypothesis. This is precisely the case with Red Hat where hypothesis, flake8 and all its dependencies are new to the distribution.

**Comment by lvh** [comment]
I'm not sure I understand. Yes, cryptography has a very extensive test suite, some of which uses optional novel testing strategies like hypothesis' generative testing. If you want to run all the tests, you're going to need all the tools the tests use, transitively.

I'm trying to figure out how to make this ticket actionable. What are you hoping to see? It sounds like you want Red Hat to package a bunch of software.

**Comment by Lukasa** [comment]
Sorry if I'm losing my mind, but it's not clear that this is a bug. What exactly is the intended action to be taken here?

Hypothesis' property-based testing is extremely valuable. I doubt cryptography will remove it, and even if they did, Hypothesis needs to be packaged at some point because [lots of projects are using it](https://hypothesis.readthedocs.org/en/latest/usage.html), including Mercurial.

The difficulty here seems to be entirely in Red Hat's internal policies. No-one is forcing Red Hat to package Hypothesis, it's available from PyPI: that's Red Hat's choice. No-one is forcing Red Hat to *test* Hypothesis: that's Red Hat's choice too. It rather seems like Red Hat has made their own bed and is now complaining about having to lie in it.

**Comment by Lukasa** [comment]
Having done some quick research, it seems that OpenStack uses flake8 in their own tests. This means there's already a Red Hat-compatible package of that library and its dependencies available. That could be used as a starting point, or you could contact the maintainer of that package who presumably works for Red Hat as part of RDO.

**Comment by npmccallum** [comment]
I am the Red Hat and Fedora maintainer of python-cryptography. I'm trying to communicate a few things and I thought this was the best place to do so.

1. Every time a dependency is added, it comes with a cost. This cost grows proportionally with the size of the dependency chain. Adding a single dependency isn't a big deal. Adding a dependency with a stack of 12 new small packages with circular dependencies is. Again, the point here isn't the single package but the cumulative cost. The actionable item here is that I'd like pyca to consider this cost and weigh it before adding new dependencies. One single glance at hypothesis' setup.py should have sent up red flags. This cost is even higher when the dependencies are runtime (versus build-time).

2. It would be nice to find a way to make, in particular, the test suite more modular. For instance, can't you just skip the tests that use hypothesis if it is not found. Having a huge test suite is great. I'm a fan. Having a huge batch of dependencies for a small number of tests isn't so great. This allows for incremental improvement of packaging (rather than having to do all the work up front).

3. The cryptography project is now dragging in large parts of a software ecosystem. As a large consumer, you have some weight to convince your dependencies to minimize their dependencies. This effort would produce a higher value to python-cryptography since the deployment cost is lowered. In particular, these discussions should happen before adopting new dependencies. For instance (just to pick on hypothesis some more), seeing the large number of dependencies in hypothesis' setup.py should start a conversation with them about modularizing or reducing their dependency footprint.

In short, I'm not just complaining about a new dependency. I'm trying to communicate the pain that new dependencies cause your downstreams and suggest some ways to minimize that pain.

**Comment by npmccallum** [comment]
@Lukasa OpenStack is not RHEL; python-cryptography is.

**Comment by Lukasa** [comment]
> One single glance at hypothesis' setup.py should have sent up red flags.

I did glance at Hypothesis setup.py when I tried to work out what package was blocking you. Hypothesis has exactly two mandatory test dependencies: py.test (presumably already packaged because Requests uses it) and flake8. Neither of these are packages that would send up the slightest warning sign to a Python developer, and both are already used by cryptography in some form or another. Why should any of the cryptography devs been nervous about that file?

> For instance (just to pick on hypothesis some more), seeing the large number of dependencies in hypothesis' setup.py should start a conversation with them about modularizing or reducing their dependency footprint.

Sorry, but again, this is perplexing to me. Hypothesis has *zero* `install_requires` dependencies. It has *two* `tests_require` dependencies, one of which is already packaged and . All the remaining dependencies are extras that cryptography does not bring in. This is *not* an onerous testing burden.

In fact, I manually walked the dependency tree for you. Here it is:

- Hypothesis
    - py.test (already packaged, will not walk it further)
    - flake8
         - pyflakes (leaf, no dependencies)
         - pep8 (leaf, no dependencies)
         - mccabe (leaf, no dependencies)
         - nose (already packaged, will not walk it further)
         - mock (probably already packaged, but let's assume it isn't)
             - pbr (leaf, no dependencies)

That is not very many dependencies, and is also 3 fewer than your count. What dependencies am I missing?

Additionally, flake8 and py.test were already used by cryptography in the tox target, and so from the perspective of the core project adding Hypothesis represented an addition of *exactly one* dependency.

Modularising the tests and removing the test dependency is a very good way of ensuring that tests are not run until the CI system is hit. Additionally, if a mistake is made in the code, a modularized test suite can subtly fail to run all the tests. If a test is worth writing, it is worth running.

> This effort would produce a higher value to python-cryptography since the deployment cost is lowered.

Test dependencies don't add to the deployment cost of cryptography.

> OpenStack is not RHEL; python-cryptography is.

Sure, but what I was indicating is that you don't have to do the packaging work from scratch: the packages already exist. You just need to go through whatever bureaucracy is required to turn them into RHEL packages.

**Comment by npmccallum** [comment]
At this point I think that no amount of communication on our part will convince you that every new release of python-cryptography has been a huge problem for us or that pyca should care in any way. Though I think it reflects poorly that when your downstreams suggest a pain point you simply respond that their policies are the problem and should be changed.

**Comment by lvh** [comment]
What?! You suggested cryptography should have seen that hypothesis would be a pain, and @Lukasa not only asked to clarify (because hypothesis looked, and still looks, like a reasonable thing to add), went through the trouble of checking what those dependencies looked like, and gave you a starting point to package the one thing that might be a problem. How does that equate to "PyCA doesn't care about your problems"?! @Lukasa is very clearly asking to understand what those problems are.

**Comment by npmccallum** [comment]
No @Lukasa spent time insinuating that I'm not a (true?) Python developer.

**Comment by Lukasa** [comment]
@npmccallum I apologise unreservedly for that implication, it was not my intention to give it.

**Comment by npmccallum** [comment]
Here is my current task list to update python-cryptography in RHEL. This is not comprehensive as I gave up after looking at traceback2.

* cryptography
  * cffi (rebase)
  * setuptools (rebase)
  * pyasn1 (rebase)
  * ipaddress (new)
  * idna (new)
  * hypothesis (new)
    * flake8 (new)
      * pyflakes (new)
      * pep8 (new)
      * mccabe (new)
      * mock (new)
        * pbr (new)
        * funcsigs (new)
        * unittest2 (new)
          * traceback2 (new)
            * unittest2 (circular)
            * contextlib2 (new)
            * linecache2 (new)

**Comment by Lukasa** [comment]
@npmccallum Aha, that's the dependency I missed: `mock`s dependency on unittest2 is in setup.cfg because it uses pbr, which I forgot to check.

That's definitely a sizeable chain of dependencies, and packaging them all is probably going to be unpleasant. If it's really not going to be possible to use the work already done for OpenStack in RDO, it may be that the better approach is to add a downstream patch that forcibly skips the Hypothesis-based tests, which should be as simple as adding `import pytest; hypothesis = pytest.importorskip('hypothesis')` at the top of the file.

With that approach, Hypothesis is run on any platform that can happily install it, and you can save yourself a huge chunk of work packaging all those dependencies.

If the cryptography developers are willing to consider changes here, another approach would be to exclude the Hypothesis-based testing from the tests run by `python setup.py test`. Another similar code-based approach would be to use py.test's `importorskip` hook.

That seems like the set of options. Ordered, from best to worst:

1. Bring the packages over from OpenStack-land.
2. Use `importorskip` patch to avoid requiring to package.
3. Make the tests optional in cryptography more generally.
4. (And a very distant 4 at that) Do all the packaging work required to make these tests pass.

**Comment by mitsuhiko** [comment]
@npmccallum if I see this correctly all those dependencies come in via flake8? In that case maybe the solution is to make flake8 optional in hypothesis. I would assume that's an entirely pointless dependency for regular tests.

**Comment by Lukasa** [comment]
I should note that Red Hat may actively not want to package Hypothesis, given the [guidelines from that project](https://hypothesis.readthedocs.org/en/latest/packaging.html).

**Comment by npmccallum** [comment]
@mitsuhiko I would be *very* open to that solution.

**Comment by npmccallum** [comment]
@Lukasa Correct. If I understand hypothesis correctly, it can also lead to entropy in the test suite which may produce issues with reproducability. Reproducing exactly the same result every time is a core value of our distribution.

**Comment by Lukasa** [comment]
@mitsuhiko @npmccallum In fact the `tests_require` in Hypothesis is a vestigial remainder of a time when `setup.py test` worked for that project, but it is not currently supported in Hypothesis anyway so Red Hat would have to write a custom testing script for that project. Given that Red Hat would be defining what the test are on that platform, I believe Red Hat can simply entirely ignore the tests_require portion of setup.py.

I suspect @DRMacIver would be open to removing flake8 from tests_require. However, please be aware that Hypothesis has ... opinions ... about being packaged downstream which may make that route less appealing than it appears to be.

**Comment by Lukasa** [comment]
See also DRMacIver/hypothesis#257.

**Comment by npmccallum** [comment]
As I see it I have three options:
1. Disable hypothesis tests altogether and remove its unused testing dependencies.
2. Make flake8 optional for hypothesis testing.
3. Make hypothesis optional for cryptography testing.

I'm currently doing (1) as a workaround. My near-term preference is for (3) since it also removes entropy from the testing process.

It may also be helpful to recognize that there are two kinds of tests: those developers want to run and those distributors want to run. I think the hypothesis tests are clearly in the first category.

**Comment by Lukasa** [comment]
> It may also be helpful to recognize that there are two kinds of tests: those developers want to run and those distributors want to run. I think the hypothesis tests are clearly in the first category.

I can understand that. @reaperhulk, do you believe the Hypothesis-based tests are required to validate that a cryptography install on a platform is functioning correctly, or does cryptography have sufficient tests of Fernet to be comfortable that it's fine if the Hypothesis tests don't run?

If you think that it'd be safe for downstream to avoid running the Hypothesis test if they're just trying to validate, I'm open to providing a patch that skips the Hypothesis based tests when `python setup.py test` is run. @npmccallum would that patch be sufficient for your purposes?

**Comment by reaperhulk** [comment]
We keep `python setup.py test` compatibility specifically because downstream packagers use it, so skipping hypothesis there isn't out of the question, but I'm concerned about the long term implications of that.

At this time the hypothesis tests are not required (and do not affect coverage of lines/branches). That said, without expanding our CI to measure coverage when these tests are excluded it will be difficult (if not impossible) to guarantee that hypothesis won't become extremely important for validation over time.

**Comment by npmccallum** [comment]
@Lukasa It might be. I'd like to better understand if hypothesis is actually inserting randomness or not (the docs are vague on this point; I may have missed something). I want to avoid a situation where tests in the build environment randomly fail.

If hypothesis is inserting entropy into the test process, I'd like it to be optional (option 3 above). If it is not inserting entropy into the test process, I'd like it to be required (option 1 above).

**Comment by npmccallum** [comment]
To be absolutely clear: I'm all for an extensive test suite even when it requires additional dependencies to do right. I want to avoid needless bloat and irreproducibility.

**Comment by Lukasa** [comment]
@npmccallum Hypothesis can be made to be deterministic, but that greatly hampers its utility for the upstream developers, where Hypothesis' ability to be "creative" is extremely useful.

An interesting question is whether, as part of the release process, we could generate a Hypothesis example database and then force Hypothesis to restrict itself to testing only those examples when the tests are executed from `python setup.py test`. Let me investigate that idea and get back to you shortly.

**Comment by sigmavirus24** [comment]
Hypothesis isn't needless bloat though. Most of the bloat in your list above is courtesy of the mock project being maintained by the testing-cabal. (That's also the group that introduced the circular dependency.) Perhaps that argument should be made to them instead of to cryptography that isn't even directly adding mock as a dependency. (This is spoken as a very occasional contributor to cryptography and the maintainer of flake8.)

**Comment by reaperhulk** [comment]
I'm not sure I follow about the desire for deterministic tests in this case. If the test suite failed intermittently that would actually be a great warning sign that the software should not be packaged, as it contains a bug that needs to be corrected. Am I mistaken?

You could, of course, have a scenario where the test failure is spurious caused by a bug in the test framework, but that is already a possibility (and indeed has happened several times in the recent past with pytest for us).

**Comment by npmccallum** [comment]
@Lukasa Great! Thanks! I'm reopening the issue since we actually have a way forward now.

**Comment by reaperhulk** [comment]
(As a side note I'd like to thank everyone involved for turning this into a productive discussion)

**Comment by npmccallum** [comment]
@reaperhulk In our case, if the failure is intermittent we might never see it in the build process but then later get Failure To Build From Source bugs from our downstreams.

@sigmavirus24 I want to be absolutely clear that I never claimed that hypothesis was needless bloat. Its flake8 dependency was, however.

**Comment by Lukasa** [comment]
Ok, so here's the state of play on deterministic Hypothesis.

It's possible to configure Hypothesis in such a way that it only runs tests given by `@example`. Currently there are none of those. One possible option here would be for the cryptography team to commit to having `python setup.py test` configure Hypothesis in that manner, and commit to having enough `@example` examples to achieve code coverage.

There are some downsides here: those examples need to be created somehow, and that would be something that the team requires for code coverage purposes. As @reaperhulk has indicated, we'd also be wanting to make sure that we validate our code coverage requirements in this case.

That's a biggish chunk of work. I'm open to doing some of the initial chunk if everyone is interested in pursuing that route, but it has ongoing costs that the team may not want to bear. It may be better to simply skip Hypothesis tests entirely downstream.

@reaperhulk? Any other core dev with an opinion?

**Comment by npmccallum** [comment]
@Lukasa I think, given the packaging guidelines from hypothesis, we should probably seek to make hypothesis optional for tests. We definitely want to run the tests. But I don't think we can commit to the update schedule hypothesis requests of us. I don't think RH will be the only entity in this position. Certainly Debian and Ubuntu LTS will be.

**Comment by Lukasa** [comment]
@npmccallum That's fair enough. In that case, I'll investigate that change.

**Comment by npmccallum** [comment]
It is really a shame given the declarative approach of hypothesis. It has a lot of benefits. I would hope making hypothesis optional could be revisited in the future.

**Comment by npmccallum** [comment]
Also, please keep in mind that I am happy to add a flag to skip hypothesis tests to our build. That way they can still be run by default in the upstream code.

**Comment by Lukasa** [comment]
@npmccallum Right now I'm investigating whether we should use an environment variable, or whether I can get python setup.py test to do all the magic we need to remove them. Ideally you'll have to maintain no patch in order to achieve this. @reaperhulk's primary concern is just that Red Hat's tests of cryptography run a sufficient amount *without* Hypothesis to validate function to his exacting standards. ;)

**Comment by npmccallum** [comment]
I think @reaperhulk's concern is an important one. I don't believe anyone here is looking to degrade the quality.

Perhaps we should also approach hypothesis about the stability question so that we can just package hypothesis?

**Comment by Lukasa** [comment]
@npmccallum The last conversation I had with @DRMacIver suggested that he didn't expect a slowdown in releases until 3.0.0, which is right now mostly hypothetical. I'd expect the current state of Hypothesis stability to remain true for the forseeable future.

**Comment by Lukasa** [comment]
Ok, for the consideration of this entire conversation: #2667.

**Comment by reaperhulk** [comment]
I think that the issues @DRMacIver has with packaging are not so much an API stability concern (after all, if you package hypothesis with cryptography then both of those are pinned so they'll keep working) so much as a source of significant headaches when users install those old packages and expect support for [[whatever issue]] despite much newer versions being available. I may be mistaken though?

For the cryptography 1.2 release the simplest solution is probably to do a downstream `importorskip` patch to the hypothesis test module (along with removing it from test_requires). While we could do a `setup.py test` patch upstream I'd prefer to go down the hypothesis packaging avenue first to avoid my previously stated long term concerns about test coverage.

**Comment by DRMacIver** [comment]
I confirm those are more or less my issues. The Hypothesis API is extremely stable - there were a few minor API breakages in 2.0, but otherwise it's pretty rigorously semantically versioned.

It's less that I'm worried about users getting support (I'll just tell them "It's fixed in version xyz but your vendor hasn't packaged that") and more that e.g. if I fix a performance issue the result will be a whole bunch of users who never get that fix because their vendor isn't bothering to update the package and will blame Hypothesis for being slow instead of their vendor for not updating. This is not a hypothetical (no pun intended) example.

**Comment by simo5** [comment]
@repearhulk if hypothesys API is stable I think it is ok to require it, the problem for a distribution though comes when 2 packages end up depending on 2 different versions of hypothesys and neither can use the other version. This is not an hypothetical (no pun as well :) and happens regulraly with fast changing libraries or modules that become quickly popular. I guess we can deal with it when it happens but having the option to disable tests in a bind is valuable.

**Comment by npmccallum** [comment]
OK, given @DRMacIver's comment and @Lukasa's patch to remove the vestigial flake8 dep in hypothesis, my currently inclination is to keep hypothesis as a required dependency. I have created a new build of hypothesis for Fedora which removes the flake8 dep and I have proposed it for inclusion. I have also requested to be a comaintainer of hypothesis in Fedora so that we can get the packages there in a better shape.

The only remaining issues so far as I can see is the question of entropy in build-time tests. But I'm okay to live with that.

**Comment by npmccallum** [comment]
@DRMacIver Are there plans to resurrect setup.py test?

**Comment by DRMacIver** [comment]
No.

**Comment by npmccallum** [comment]
Okay. Thanks everyone! I think we can call this closed.

**Comment by Lukasa** [comment]
\o/ Thanks all!
